File: 011_subdivision/solution/reverse_subdivision.m
function u = reverse_subdivision(V,F,uu,k)
%REVERSE_SUBDIVISION Given a function uu on a mesh which has been
%subdivided k times from the coarse V,F, reconstruct a function u on the
%coarse mesh V,F.
%
% u = reverse_subdivision(V,F,uu,k);
%
% Inputs:
%  V,F  the coarse input mesh
%  uu  a function on the subdivided mesh loop(V,F,k)
%  k  how many times to subdivide
% Outputs:
%  u  a proxy for the function uu on the coarse mesh V,F
%

[~,~,S] = loop(V,F,k);
u = (S'*S) \ (S'*uu);

end

File: 011_subdivision/solution/reverse_subdivision.py
import gpytoolbox as gpy, scipy as sp

def reverse_subdivision(V, F, uu, k):
    """Given a function uu on a mesh which has been subdivided k times from the
    coarse V,F, reconstruct a function u on the coarse meth V,F.
    """

    _,_,S = gpy.subdivide(V,F, iters=k, method='loop', return_matrix=True)
    u = sp.sparse.linalg.spsolve(S.T * S, S.T * uu)

    return u
